tests/gasless_test.go

```diff
@@ -0,0 +1,401 @@
+		routerAddr:        routerAddr,
+		routerContract:    routerContract,
+		gsrAddr:           gsrAddr,
+		gsrContract:       gsrContract,
+	}
+	setupLiquidity(t, owner, contracts, chain)
+
+	/* ------------------------------------ Main test process ------------------------------------- */
+	// get amounts out
+	swapAmmount := new(big.Int).Mul(big.NewInt(1), bigKaia)
+	amountsOut, err := routerContract.GetAmountsOut(&bind.CallOpts{}, swapAmmount, []common.Address{testTokenAddr, wkaiaAddr})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		gasPriceBN         = new(big.Int).Mul(big.NewInt(50), bigGkei)
+		R1                 = new(big.Int).Mul(big.NewInt(21000), gasPriceBN)
+		R2                 = new(big.Int).Mul(big.NewInt(100000), gasPriceBN)
+		R3                 = new(big.Int).Mul(big.NewInt(500000), gasPriceBN)
+		ammontRepay        = new(big.Int).Add(R1, new(big.Int).Add(R2, R3))
+		transferToken      = new(big.Int).Mul(big.NewInt(100), bigKaia)
+		swapExpectedOutput = amountsOut[1]
+		margin             = new(big.Int).Div(swapExpectedOutput, big.NewInt(100))
+		minAmountOut       = new(big.Int).Add(ammontRepay, margin)
+	)
+
+	// transfer test token
+	testTokenTransferTx, err := testTokenContract.Transfer(bind.NewKeyedTransactor(owner.Keys[0]), accounts[0].Addr, transferToken)
+	if err != nil {
+		t.Fatal(err)
+	}
+	testTokenTransferReceipt := waitReceipt(chain, testTokenTransferTx.Hash())
+	if testTokenTransferReceipt == nil {
+		t.Fatal("failed to transfer test token")
+	}
+	owner.Nonce += 1
+
+	balanceOfTestAcc, _ := testTokenContract.BalanceOf(&bind.CallOpts{}, accounts[0].Addr)
+	balanceOfOwner, _ := testTokenContract.BalanceOf(&bind.CallOpts{}, owner.Addr)
+	t.Log("test acc balance: ", balanceOfTestAcc)
+	t.Log("owner balance: ", balanceOfOwner)
+
+	// send approveTx
+	optsForApprove := bind.NewKeyedTransactor(accounts[0].Keys[0])
+	optsForApprove.GasLimit = 300000
+	optsForApprove.Nonce = big.NewInt(int64(accounts[0].Nonce))
+	approveTx, err := testTokenContract.Approve(optsForApprove, gsrAddr, swapAmmount)
+	if err != nil {
+		t.Fatal(err)
+	}
+	accounts[0].Nonce += 1
+	t.Log("approveTxHash", approveTx.Hash().Hex())
+
+	// send swapTx
+	optsForSwap := bind.NewKeyedTransactor(accounts[0].Keys[0])
+	optsForSwap.GasLimit = 300000
+	optsForSwap.Nonce = big.NewInt(int64(accounts[0].Nonce))
+	swapTx, err := gsrContract.SwapForGas(optsForSwap, testTokenAddr, swapAmmount, minAmountOut, ammontRepay)
+	if err != nil {
+		t.Fatal(err)
+	}
+	accounts[0].Nonce += 1
+	t.Log("swapTxHash", swapTx.Hash().Hex())
+
+	// check if account[0] without kaia can send tx
+	approveTxReceipt := waitReceipt(chain, approveTx.Hash())
+	require.NotNil(t, approveTxReceipt)
+	require.Equal(t, types.ReceiptStatusSuccessful, approveTxReceipt.Status, "approveTx failed")
+
+	swapTxReceipt := waitReceipt(chain, swapTx.Hash())
+	require.NotNil(t, swapTxReceipt)
+	require.Equal(t, types.ReceiptStatusSuccessful, swapTxReceipt.Status, "swapTx failed")
```

Comment:
Why not do you check the amount of $KAIA and TEST token of accounts[0] before and after gassless process?